File: AddressBiulder/src/AddressBuilder.kt
class AddressBuilder {
    private val address = Address()

    fun setCountry(country: String): AddressBuilder {
        address.country = country
        return this
    }

    fun setSubject(type: SubjectType, name: String): AddressBuilder {
        address.subject = SubjectInfo(type, name)
        return this
    }

    fun setMunicipalDistrict(district: String): AddressBuilder {
        address.municipalDistrict = district
        return this
    }

    fun setSettlement(settlement: String): AddressBuilder {
        address.settlement = settlement
        return this
    }

    fun setLocality(locality: String): AddressBuilder {
        address.locality = locality
        return this
    }

    fun setPlanningStructure(structure: String): AddressBuilder {
        address.planningStructure = structure
        return this
    }

    fun setRoad(type: RoadType, name: String): AddressBuilder {
        address.roadNetwork = RoadInfo(type, name)
        return this
    }

    fun setBuilding(type: BuildingType, number: String): AddressBuilder {
        address.addressObjectType = AddressObjectType.Building
        address.building = BuildingInfo(type, number)
        return this
    }

    fun setLandParcel(type: LandParcelType, number: String): AddressBuilder {
        address.addressObjectType = AddressObjectType.LandParcel
        address.landParcel = LandParcelInfo(type, number)
        return this
    }

    fun setParkingSpace(type: ParkingSpaceType, number: String): AddressBuilder {
        address.addressObjectType = AddressObjectType.ParkingSpace
        address.parkingSpace = ParkingSpaceInfo(type, number)
        return this
    }

    fun setPremise(type: PremiseType, number: String): AddressBuilder {
        address.addressObjectType = AddressObjectType.Premise
        address.premise = PremiseInfo(type, number)
        return this
    }

    fun build(): String {
        return address.generateFullAddress()
    }

    fun getAddress(): Address {
        return address
    }

    fun reset() {
        address.country = ""
        address.subject = null
        address.municipalDistrict = null
        address.settlement = null
        address.locality = null
        address.planningStructure = null
        address.roadNetwork = null
        address.addressObjectType = null
        address.building = null
        address.landParcel = null
        address.premise = null
        address.parkingSpace = null
    }
}

File: AddressBuilding/AddressBuilder.cs
namespace AddressBuilding;


public class AddressBuilder
{
    private Address _address = new Address();

    public AddressBuilder SetCountry(string country)
    {
        _address.Country = country;
        return this;
    }

    public AddressBuilder SetSubject(SubjectType type, string name)
    {
        _address.Subject = new SubjectInfo(type, name);
        return this;
    }

    public AddressBuilder SetMunicipalDistrict(string district)
    {
        _address.MunicipalDistrict = district;
        return this;
    }

    public AddressBuilder SetSettlement(string settlement)
    {
        _address.Settlement = settlement;
        return this;
    }

    public AddressBuilder SetLocality(string locality)
    {
        _address.Locality = locality;
        return this;
    }

    public AddressBuilder SetPlanningStructure(string structure)
    {
        _address.PlanningStructure = structure;
        return this;
    }

    public AddressBuilder SetRoad(RoadType type, string name)
    {
        _address.RoadNetwork = new RoadInfo(type, name);
        return this;
    }

    public AddressBuilder SetBuilding(BuildingType type, string number)
    {
        _address.AddressObjectType = AddressObjectType.Building;
        _address.Building = new BuildingInfo(type, number);
        return this;
    }

    public AddressBuilder SetLandParcel(LandParcelType type, string number)
    {
        _address.AddressObjectType = AddressObjectType.LandParcel;
        _address.LandParcel = new LandParcelInfo(type, number);
        return this;
    }

    public AddressBuilder SetParkingSpace(ParkingSpaceType type, string number)
    {
        _address.AddressObjectType = AddressObjectType.ParkingSpace;
        _address.ParkingSpace = new ParkingSpaceInfo(type, number);
        return this;
    }

    public AddressBuilder SetPremise(PremiseType type, string number)
    {
        _address.AddressObjectType = AddressObjectType.Premise;
        _address.Premise = new PremiseInfo(type, number);
        return this;
    }

    public string Build()
    {
        return _address.GenerateFullAddress();
    }
}